In [1]:
!pip install psycopg2

Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.


# Import the required libraries

In [3]:
import psycopg2
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
from IPython.core.display import Image, display

# Connecting to the dvdrental database

In [5]:
#Connecting to the server
conn = psycopg2.connect(database='dvdrental', user='postgres', password='post')

**DVD Rental ER Diagram**

<img src="dvdrental_diagram.png" alt="Drawing" style="width: 500px;"/>

**Contents of each table**

In [7]:
sql = ["SELECT * from actor","SELECT * from actor_info","SELECT * from address",
       "SELECT * from category","SELECT * from FILM","SELECT * from city","SELECT * from film_category",
      "SELECT * from inventory","SELECT * from customer","SELECT * from rental",
      "SELECT * from payment","SELECT * from language","SELECT * from staff",
      "SELECT * from film_actor","SELECT * from country","SELECT * from store"]


for sql in sql:
    print(sql.lstrip("SELECT * from ").upper()+" TABLE")
    pd.read_sql_query(sql, conn).head(5)
    print('\n')

ACTOR TABLE


actor_id first_name     last_name             last_update
0         1   Penelope       Guiness 2013-05-26 14:47:57.620
1         2       Nick      Wahlberg 2013-05-26 14:47:57.620
2         3         Ed         Chase 2013-05-26 14:47:57.620
3         4   Jennifer         Davis 2013-05-26 14:47:57.620
4         5     Johnny  Lollobrigida 2013-05-26 14:47:57.620



ACTOR_INFO TABLE


actor_id first_name     last_name  \
0         1   Penelope       Guiness   
1         2       Nick      Wahlberg   
2         3         Ed         Chase   
3         4   Jennifer         Davis   
4         5     Johnny  Lollobrigida   

                                           film_info  
0  Animation: Anaconda Confessions, Children: Lan...  
1  Action: Bull Shawshank, Animation: Fight Jawbr...  
2  Action: Caddyshack Jedi, Forrest Sons, Classic...  
3  Action: Barefoot Manchurian, Animation: Anacon...  
4  Action: Amadeus Holy, Grail Frankenstein, Ring...



ADDRESS TABLE


address_id               address address2  district  city_id postal_code  \
0           1     47 MySakila Drive     None   Alberta      300               
1           2    28 MySQL Boulevard     None       QLD      576               
2           3     23 Workhaven Lane     None   Alberta      300               
3           4  1411 Lillydale Drive     None       QLD      576               
4           5        1913 Hanoi Way           Nagasaki      463       35200   

         phone         last_update  
0              2006-02-15 09:45:30  
1              2006-02-15 09:45:30  
2  14033335568 2006-02-15 09:45:30  
3   6172235589 2006-02-15 09:45:30  
4  28303384290 2006-02-15 09:45:30



CATEGORY TABLE


category_id       name         last_update
0            1     Action 2006-02-15 09:46:27
1            2  Animation 2006-02-15 09:46:27
2            3   Children 2006-02-15 09:46:27
3            4   Classics 2006-02-15 09:46:27
4            5     Comedy 2006-02-15 09:46:27



FILM TABLE


film_id              title  \
0      133    Chamber Italian   
1      384   Grosse Wonderful   
2        8    Airport Pollock   
3       98  Bright Encounters   
4        1   Academy Dinosaur   

                                         description  release_year  \
0  A Fateful Reflection of a Moose And a Husband ...          2006   
1  A Epic Drama of a Cat And a Explorer who must ...          2006   
2  A Epic Tale of a Moose And a Girl who must Con...          2006   
3  A Fateful Yarn of a Lumberjack And a Feminist ...          2006   
4  A Epic Drama of a Feminist And a Mad Scientist...          2006   

   language_id  rental_duration  rental_rate  length  replacement_cost rating  \
0            1                7         4.99     117             14.99  NC-17   
1            1                5         4.99      49             19.99      R   
2            1                6         4.99      54             15.99      R   
3            1                4         4.99      73             12.99  PG-13   
4            1                6         0.99      86             20.99     PG   

              last_update                     special_features  \
0 2013-05-26 14:50:58.951                           [Trailers]   
1 2013-05-26 14:50:58.951                  [Behind the Scenes]   
2 2013-05-26 14:50:58.951                           [Trailers]   
3 2013-05-26 14:50:58.951                           [Trailers]   
4 2013-05-26 14:50:58.951  [Deleted Scenes, Behind the Scenes]   

                                            fulltext  
0  'chamber':1 'fate':4 'husband':11 'italian':2 ...  
1  'australia':18 'cat':8 'drama':5 'epic':4 'exp...  
2  'airport':1 'ancient':18 'confront':14 'epic':...  
3  'boat':20 'bright':1 'conquer':14 'encount':2 ...  
4  'academi':1 'battl':15 'canadian':20 'dinosaur...



CITY TABLE


city_id                city  country_id         last_update
0        1  A Corua (La Corua)          87 2006-02-15 09:45:25
1        2                Abha          82 2006-02-15 09:45:25
2        3           Abu Dhabi         101 2006-02-15 09:45:25
3        4                Acua          60 2006-02-15 09:45:25
4        5               Adana          97 2006-02-15 09:45:25



ILM_CATEGORY TABLE


film_id  category_id         last_update
0        1            6 2006-02-15 10:07:09
1        2           11 2006-02-15 10:07:09
2        3            6 2006-02-15 10:07:09
3        4           11 2006-02-15 10:07:09
4        5            8 2006-02-15 10:07:09



INVENTORY TABLE


inventory_id  film_id  store_id         last_update
0             1        1         1 2006-02-15 10:09:17
1             2        1         1 2006-02-15 10:09:17
2             3        1         1 2006-02-15 10:09:17
3             4        1         1 2006-02-15 10:09:17
4             5        1         2 2006-02-15 10:09:17



CUSTOMER TABLE


customer_id  store_id first_name last_name  \
0          524         1      Jared       Ely   
1            1         1       Mary     Smith   
2            2         1   Patricia   Johnson   
3            3         1      Linda  Williams   
4            4         2    Barbara     Jones   

                                 email  address_id  activebool create_date  \
0         jared.ely@sakilacustomer.org         530        True  2006-02-14   
1        mary.smith@sakilacustomer.org           5        True  2006-02-14   
2  patricia.johnson@sakilacustomer.org           6        True  2006-02-14   
3    linda.williams@sakilacustomer.org           7        True  2006-02-14   
4     barbara.jones@sakilacustomer.org           8        True  2006-02-14   

              last_update  active  
0 2013-05-26 14:49:45.738       1  
1 2013-05-26 14:49:45.738       1  
2 2013-05-26 14:49:45.738       1  
3 2013-05-26 14:49:45.738       1  
4 2013-05-26 14:49:45.738       1



ENTAL TABLE


rental_id         rental_date  inventory_id  customer_id  \
0          2 2005-05-24 22:54:33          1525          459   
1          3 2005-05-24 23:03:39          1711          408   
2          4 2005-05-24 23:04:41          2452          333   
3          5 2005-05-24 23:05:21          2079          222   
4          6 2005-05-24 23:08:07          2792          549   

          return_date  staff_id         last_update  
0 2005-05-28 19:40:33         1 2006-02-16 02:30:53  
1 2005-06-01 22:12:39         1 2006-02-16 02:30:53  
2 2005-06-03 01:43:41         2 2006-02-16 02:30:53  
3 2005-06-02 04:33:21         1 2006-02-16 02:30:53  
4 2005-05-27 01:32:07         1 2006-02-16 02:30:53



PAYMENT TABLE


payment_id  customer_id  staff_id  rental_id  amount  \
0       17503          341         2       1520    7.99   
1       17504          341         1       1778    1.99   
2       17505          341         1       1849    7.99   
3       17506          341         2       2829    2.99   
4       17507          341         2       3130    7.99   

                payment_date  
0 2007-02-15 22:25:46.996577  
1 2007-02-16 17:23:14.996577  
2 2007-02-16 22:41:45.996577  
3 2007-02-19 19:39:56.996577  
4 2007-02-20 17:31:48.996577



LANGUAGE TABLE


language_id                  name         last_update
0            1  English              2006-02-15 10:02:19
1            2  Italian              2006-02-15 10:02:19
2            3  Japanese             2006-02-15 10:02:19
3            4  Mandarin             2006-02-15 10:02:19
4            5  French               2006-02-15 10:02:19



STAFF TABLE


staff_id first_name last_name  address_id                         email  \
0         1       Mike   Hillyer           3  Mike.Hillyer@sakilastaff.com   
1         2        Jon  Stephens           4  Jon.Stephens@sakilastaff.com   

   store_id  active username                                  password  \
0         1    True     Mike  8cb2237d0679ca88db6464eac60da96345513964   
1         2    True      Jon  8cb2237d0679ca88db6464eac60da96345513964   

                 last_update  \
0 2006-05-16 16:13:11.793280   
1 2006-05-16 16:13:11.793280   

                                             picture  
0  [b'\x89', b'P', b'N', b'G', b'\r', b'\n', b'Z'...  
1                                               None



ILM_ACTOR TABLE


actor_id  film_id         last_update
0         1        1 2006-02-15 10:05:03
1         1       23 2006-02-15 10:05:03
2         1       25 2006-02-15 10:05:03
3         1      106 2006-02-15 10:05:03
4         1      140 2006-02-15 10:05:03



COUNTRY TABLE


country_id         country         last_update
0           1     Afghanistan 2006-02-15 09:44:00
1           2         Algeria 2006-02-15 09:44:00
2           3  American Samoa 2006-02-15 09:44:00
3           4          Angola 2006-02-15 09:44:00
4           5        Anguilla 2006-02-15 09:44:00



STORE TABLE


store_id  manager_staff_id  address_id         last_update
0         1                 1           1 2006-02-15 09:57:12
1         2                 2           2 2006-02-15 09:57:12

**Datatypes of each table**

In [8]:
val = ['actor','actor_info','address','category','city','film_category','inventory',
      'customer','rental','payment','language','staff','film_actor','film','country','store']

for val in val:
    sql = '''select column_name, data_type from information_schema.columns
    where table_name = '{}'
    '''.format(val)
    print('{} Table'.format(val.upper()))
    pd.read_sql_query(sql,conn)

ACTOR Table


column_name                    data_type
0     actor_id                      integer
1   first_name            character varying
2    last_name            character varying
3  last_update  timestamp without time zone

ACTOR_INFO Table


column_name          data_type
0    actor_id            integer
1  first_name  character varying
2   last_name  character varying
3   film_info               text

ADDRESS Table


column_name                    data_type
0   address_id                      integer
1      address            character varying
2     address2            character varying
3     district            character varying
4      city_id                     smallint
5  postal_code            character varying
6        phone            character varying
7  last_update  timestamp without time zone

CATEGORY Table


column_name                    data_type
0  category_id                      integer
1         name            character varying
2  last_update  timestamp without time zone

CITY Table


column_name                    data_type
0      city_id                      integer
1         city            character varying
2   country_id                     smallint
3  last_update  timestamp without time zone

FILM_CATEGORY Table


column_name                    data_type
0      film_id                     smallint
1  category_id                     smallint
2  last_update  timestamp without time zone

INVENTORY Table


column_name                    data_type
0  inventory_id                      integer
1       film_id                     smallint
2      store_id                     smallint
3   last_update  timestamp without time zone

CUSTOMER Table


column_name                    data_type
0  customer_id                      integer
1     store_id                     smallint
2   first_name            character varying
3    last_name            character varying
4        email            character varying
5   address_id                     smallint
6   activebool                      boolean
7  create_date                         date
8  last_update  timestamp without time zone
9       active                      integer

RENTAL Table


column_name                    data_type
0     rental_id                      integer
1   rental_date  timestamp without time zone
2  inventory_id                      integer
3   customer_id                     smallint
4   return_date  timestamp without time zone
5      staff_id                     smallint
6   last_update  timestamp without time zone

PAYMENT Table


column_name                    data_type
0    payment_id                      integer
1   customer_id                     smallint
2      staff_id                     smallint
3     rental_id                      integer
4        amount                      numeric
5  payment_date  timestamp without time zone

LANGUAGE Table


column_name                    data_type
0  language_id                      integer
1         name                    character
2  last_update  timestamp without time zone

STAFF Table


column_name                    data_type
0      staff_id                      integer
1    first_name            character varying
2     last_name            character varying
3    address_id                     smallint
4         email            character varying
5      store_id                     smallint
6        active                      boolean
7      username            character varying
8      password            character varying
9   last_update  timestamp without time zone
10      picture                        bytea

FILM_ACTOR Table


column_name                    data_type
0     actor_id                     smallint
1      film_id                     smallint
2  last_update  timestamp without time zone

FILM Table


column_name                    data_type
0            film_id                      integer
1              title            character varying
2        description                         text
3       release_year                      integer
4        language_id                     smallint
5    rental_duration                     smallint
6        rental_rate                      numeric
7             length                     smallint
8   replacement_cost                      numeric
9             rating                 USER-DEFINED
10       last_update  timestamp without time zone
11  special_features                        ARRAY
12          fulltext                     tsvector

COUNTRY Table


column_name                    data_type
0   country_id                      integer
1      country            character varying
2  last_update  timestamp without time zone

STORE Table


column_name                    data_type
0          store_id                      integer
1  manager_staff_id                     smallint
2        address_id                     smallint
3       last_update  timestamp without time zone

# SQL Queries Practice Questions

# 1. Basic Select

<img src="city_1.png" alt="Drawing" style="width: 250px;"/>

**Q1. Query all columns for all cities with country_id equal to 20.**

In [ ]:
sql =''' SELECT * from city
WHERE country_id = 20'''

pd.read_sql_query(sql, conn).head(5)

**Q2. Query the names for all cities with country_id equal to 40.**

In [ ]:
sql ='''

'''

pd.read_sql_query(sql, conn).head(10)

**Q3. Query a list of first_name and last_name from actor table**

<img src="actor_1.png" alt="Drawing" style="width: 250px;"/>

In [ ]:
sql ='''  

'''

pd.read_sql_query(sql, conn).head(10)

**Q4. Query a list of city names and city ids from CITY with even ID numbers only. You may print the 
results in any order, but must exclude duplicates from your answer**

<img src="city_1.png" alt="Drawing" style="width: 250px;"/>

In [ ]:
sql ='''

'''

pd.read_sql_query(sql, conn).head(10)

**Q5. Let  _n_  be the number of country_id entries in CITY, and let  _m_  be the number of distinct country_id in CITY; query the value of  _n-m_  from CITY. In other words, find the difference between the total number of country_id entries in the table and the number of distinct country_id entries in the table.**

In [ ]:
sql =''' 

'''

pd.read_sql_query(sql, conn).head(10)

**Q6. Query the two cities in CITY with the shortest and longest city 
names, as well as their respective lengths (i.e.: number of characters in the name). 
If there is more than one smallest or largest city, choose the one that comes first when ordered alphabetically.**

NOTE: Use **_UNION ALL_** command if you want to combine more than one query


In [ ]:
sql ='''

'''

pd.read_sql_query(sql, conn).head(10)

**Q7. Query the list of city names starting with vowels (i.e., a, e, i, o, or u) from CITY. 
Your result cannot contain duplicates.**

In [ ]:
sql =''' 

'''

pd.read_sql_query(sql, conn).head(10)

**Q8. Query the list of city names ending with vowels (i.e., a, e, i, o, or u) from CITY. 
Your result cannot contain duplicates.**

In [ ]:
sql =''' 

'''

pd.read_sql_query(sql, conn).head(10)

**Q9. Query the list of city names from CITY which have vowels (i.e., a, e, i, o, and u) as both their first and last characters. 
Your result cannot contain duplicates.**

In [ ]:
sql =''' 

'''

pd.read_sql_query(sql, conn).head(10)

**Q10. Query the list of city names from CITY that do not start with vowels. 
Your result cannot contain duplicates.**

In [ ]:
sql = """ 

"""

pd.read_sql_query(sql, conn).head(10)

**Q11. Query the list of city names from CITY that do end with vowels. 
Your result cannot contain duplicates.**

In [ ]:
sql = """ 

"""

pd.read_sql_query(sql, conn).head(10)

**Q12. Query the list of city names from CITY that either do not start with vowels or do not end with vowels. 
Your result cannot contain duplicates.**

In [ ]:
sql =''' 

'''

pd.read_sql_query(sql, conn).head(10)

**Q13. Query the city names in CITY whose city_id is less than 100. Order your output by the last three characters of each city name. If two or more cities both have names ending in the same last three characters (i.e.: Ahmadnagar, Bhavnagar, etc.),
secondary sort them by ascending city_id.**

In [ ]:
sql =''' 

'''

pd.read_sql_query(sql, conn).head(10)

**Q14. Write a query that prints a list of actor first names and last names from the ACTOR table sorted by first name alphabetically.**

<img src="actor_1.png" alt="Drawing" style="width: 250px;"/>

In [ ]:
sql =''' 

'''

pd.read_sql_query(sql, conn).head(10)

**Q15. Write a query that prints a list of actor first names and last names from the ACTOR table sorted by last_name alphabetically.**

In [ ]:
sql =''' 

'''

pd.read_sql_query(sql, conn).head(10)

# 2. Advanced Select

**Q1. Write a query identifying the type of each record in the PAYMENT table using its amount column. Output one of the following statements for each record in the table:**

**High_sale**: amount > 11
<br>
**Medium_sale**:   5 > amount > 11
<br>
**Low_sale**:    0 > amount > 5
<br>
**Free_sale**:   amount = 0

<img src="payment.png" alt="Drawing" style="width: 250px;"/>

In [ ]:
sql =''' 

'''
pd.read_sql_query(sql, conn).head(10)
pd.read_sql_query(sql, conn)['case'].value_counts()

**Q2. Generate the following two result sets:**

**1. Query an ordered list of city names in CITY, immediately followed by the country_id of each city as a parenthical(i.e.: enclosed in parentheses). For example: Abu Dhabi(101), Abha(82), Islamabad(21).**
<br>

**2. Query the number of cities in each country_id.Sort the occurrences in ascending order, and output them in the following format:**

**_"There are a total of [city_count] in country [country_id]."_**

where **[city_count]** is the number of cities in a **[country_id]**. If more than one country has the same **[city_count]**, they should be ordered by **[country_id]**.

**Limit to 5 rows for each query!**

<img src="city_1.png" alt="Drawing" style="width: 250px;"/>



**OUTPUT**
<BR>
<img src="q2_2.png" alt="Drawing" align='left' style="width: 250px;"/>

In [ ]:
sql=''' 


'''

pd.read_sql_query(sql,conn).head(10)

**Q3. Pivot the staff_id column in PAYMENT such that the average amount processed by each staff_id for customer_id = 341 and customer_id =342 is displayed.**

Hint: Use _**crosstab**_ function for pivoting. Create the _tablefunc_ extension only once.

<img src="payment.png" alt="Drawing" style="width: 250px;"/><img src="payment_dtype.png" alt="Drawing" style="width: 250px;"/>


In [ ]:
sql = '''

'''

pd.read_sql_query(sql,conn)

**Q4. Given the table schemas below, write a query to print the film_id, title, total number of actors, total inventory units. Order your output by ascending film_id.**

<img src="film.png" alt="Drawing" style="width: 150px;"/>
<img src="film_actor.png" alt="Drawing" style="width: 150px;"/>
<img src="inventory.png" alt="Drawing" style="width: 150px;"/>

**Sample Output**
<br>
<img src="sample_input_q4.png" alt="Drawing" align='left' style="width: 350px;"/>

In [ ]:
sql ='''

'''

pd.read_sql_query(sql,conn).head(10)

# 3. Aggregation


**Q1. Query a count of the number of titles in FILM with rental_duration greater than 5.**

<img src="film.png" alt="Drawing" style="width: 150px;"/>
<img src="film_contents.png" alt="Drawing" style="width: 800px;"/>

In [ ]:
sql = ''' 

'''

pd.read_sql_query(sql,conn).head(10)

**Q2. Query the total sum of rental_duration in FILM when length is less than 50.**

In [ ]:
sql = ''' 

'''

pd.read_sql_query(sql,conn)

**Q3. Query the average rental_duration in FILM when length is less than 100.**

In [ ]:
sql = ''' 

'''

pd.read_sql_query(sql,conn)

**Q4. Query the average rental_duration for all titles in FILM, rounded down to the nearest integer.**

In [ ]:
sql = '''

'''

pd.read_sql_query(sql,conn)

**Q5. Query the average of the rental_rate for all Restricted film titles in FILM. The rating identifier for Restricted is R. Round to 2 decimal places.**

In [ ]:
sql = '''

'''

pd.read_sql_query(sql,conn)

**Q6. Query the difference between the maximum and minimum length of film in FILM.**

In [ ]:
sql = '''

'''

pd.read_sql_query(sql,conn)

**Q7. Rahul was tasked with calculating the average replacement_cost for all films in the FILM table, but did not realize his keyboard's '1' key was broken until after completing the calculation. He wants your help finding the difference between his miscalculation (using replacement_cost with any removed ones), and the actual average replacement_cost.**

Write a query calculating the amount of error **(i.e _actual average replacement cost_ - miscalculated average replacement cost)**, and round it up to the next integer.

<img src="film_datatypes.png" alt="Drawing" align='left' style="width: 275px;"/>

**ACTUAL VALUES**
<br>
<img src="q7_actual.png" alt="Drawing" align='left' style="width: 375px;"/>


**ERROR VALUES**
<br>
<img src="q7_error1.png" alt="Drawing" align='left' style="width: 375px;"/>

**SAMPLE OUTPUT**
<br>
<img src="error.png" alt="Drawing" align='left' style="width: 100px;"/>

In [ ]:
sql = '''

'''

pd.read_sql_query(sql,conn).head(10)

**Q8. We define a film's rental cost per day to be (rental_rate / rental_duration) , and the maximum total rental cost per day to be the maximum total rental cost per day for any film in the FILM table.**

Write a query to find the **maximum total rental cost per day** as well as the **total number of films which have maximum total rental cost per day.**

**Note:** Round to 2 decimal places

**SAMPLE INPUT**
<br>
<img src="q8_input1.png" alt="Drawing" align='left' style="width: 375px;"/>

**SAMPLE OUTPUT**
<br>
<img src="q8_output.png" alt="Drawing" align='left' style="width: 250px;"/>

In [ ]:
sql = ''' 
          
'''

pd.read_sql_query(sql,conn)

**Q9. Query the following two values from the FILM table:**

1. The average of all values in rental_rate rounded to a scale of 2 decimal places.
2. The average of all values in replacement_cost rounded to a scale of 2 decimal places.

**Input Format**

The FILM table is described as follows:

<img src="film.png" alt="Drawing" style="width: 150px;"/>

In [ ]:
sql = ''' 
          
'''

pd.read_sql_query(sql,conn)

**Q10. Query the sum of length of film (length) from FILM having values greater than 30  and less than 50.**

In [ ]:
sql = ''' 
          
      '''

pd.read_sql_query(sql,conn)

**Q11. Query the longest value of length of film(length) from FILM that is less than 80.**

In [ ]:
sql = '''

      '''

pd.read_sql_query(sql,conn)

**Q12. Query the shortest value of length of film(length) from FILM that is less than 50.**

In [ ]:
sql = ''' 

      '''

pd.read_sql_query(sql,conn)

**Q13. Query the replacement_cost for the longest length of film (length) in STATION that is less than 90 . Round your answer to 1 decimal place.**

In [ ]:
sql = ''' 


      '''

pd.read_sql_query(sql,conn)

**Q14. Query the replacement_cost for the shortest length of film (length) in STATION that is greater than 130 . Round your answer to 1 decimal place.**

In [ ]:
sql = ''' 

      '''

pd.read_sql_query(sql,conn)

**Q15.** Consider **W(_a_,_b_)**  and **Z(_c_,_d_)** to be two points on a 2D plane.

1. **_a_** happens to equal the minimum value in rental_rate in FILM
2. **_b_** happens to equal the minimum value in replacement_cost in FILM
3. **_c_** happens to equal the maximum value in rental_rate in FILM
4. **_d_** happens to equal the maximum value in replacement_cost in FILM

Query the **Manhattan Distance** between points W and Z

In [ ]:
sql = ''' 

      '''

pd.read_sql_query(sql,conn)

**Q16.** Consider **W(_a_,_b_)**  and **Z(_c_,_d_)** to be two points on a 2D plane.

1. **_a_** happens to equal the minimum value in rental_rate in FILM
2. **_b_** happens to equal the minimum value in replacement_cost in FILM
3. **_c_** happens to equal the maximum value in rental_rate in FILM
4. **_d_** happens to equal the maximum value in replacement_cost in FILM

Query the **Euclidean Distance** between points W and Z. Round to 3 decimal places.

In [ ]:
sql = ''' 

      '''

pd.read_sql_query(sql,conn)

**Q17. A median is defined as a number separating the higher half of a data set from the lower half. Query the median of the length of film (length) from FILM.**

In [ ]:
sql = ''' 

      '''

pd.read_sql_query(sql,conn)

**Q18. A mode is defined as a number that appears most often in a set of data values. Query the mode of the length of film (length) from FILM.**

In [ ]:
sql = ''' 

      '''

pd.read_sql_query(sql,conn)

# 4. Joins

**Q1. Given the CITY and COUNTRY tables, query the names of all cities where the country is 'India'.**

Note: CITY.country_id and COUNTRY.country_id are matching key columns.

<img src="city_country.png" alt="Drawing" style="width: 250px;"/>

In [ ]:
sql = ''' 

            
      '''

pd.read_sql_query(sql,conn).head(10)

**Q2. Given the CITY and COUNTRY tables, query the names of all the country (COUNTRY.country) and their respective number of cities (CITY.city).**

Note: CITY.country_id and COUNTRY.country_id are matching key columns.

In [ ]:
sql = ''' 

            
      '''

pd.read_sql_query(sql,conn).head(10)

**Q3. Rahul was asked to create a report containing the top 10 films with the highest revenue. The report should not contain any NULL values and must have the following 3 columns: _film id, film title, total amount_.**
<br>
Order in descending order of total amount. If more than one film have the same total_amount, order alphabetically.

Note: **INVENTORY.inventory_id** and **RENTAL.inventory_id** are matching key columns.
      **RENTAL.rental_id** and **PAYMENT.rental_id** are matching key columns.

<img src="film.png" alt="Drawing" align='left' style="width: 150px;"/>
<img src="inv_ren_pay.png" alt="Drawing" align='middle' style="width: 150px;"/>

In [ ]:
sql=''' 

    '''

pd.read_sql_query(sql,conn)

**Q4. Rahul was again asked to create a report containing the top 10 films, but this time with the least revenue. The report should not contain any NULL values and must have the following 3 columns: _film id, film title, total amount_.**
<br>
Order in ascending order of total amount. If more than one film have the same total_amount, order alphabetically.

Note: **INVENTORY.inventory_id** and **RENTAL.inventory_id** are matching key columns.
      **RENTAL.rental_id** and **PAYMENT.rental_id** are matching key columns.

In [ ]:
sql=''' 

    '''

pd.read_sql_query(sql,conn)

In [ ]:
#Close the connection to database
conn.close()